### PR TITLE
Power/product metrics wrap other metrics

### LINKED ISSUE #125: Should PowerMetric store a metric inside?
I just wanted to write a small test for `det_local_metric` on a `MetricManifold{PowerManifold,PowerMetric}` , seee https://github.com/JuliaManifolds/Manifolds.jl/blob/161014340e55ef5cb1e9df00a28655209a18128a/src/manifolds/PowerManifold.jl#L240 and encountered a problem: The current version can't work, since the inner calls are never `MetricManifold`s.

In order for them to be, one has to create a `MatricManifold` of the powers internal `Manifold` and an according `Metric`.  But at that point we only have the `PowerMetric`. So I would like to propose to store a `Metric` inside the `PowerMetric` for exactely cases like this. What do you think? Like for the one manifold, one metric should be enough.

### PR DESCRIPTION
**Pull request by mateuszbaran**
This fixes #125.

### CODE REVIEW

File: src/manifolds/PowerManifold.jl [hunk 1]
```diff
@@ -82,13 +82,17 @@ function PowerManifold(
 end
 
 @doc raw"""
-    PowerMetric <: Metric
+    PowerMetric(metric::metric)
```
Comment (kellertuer) [inline src/manifolds/PowerManifold.jl:85]:
```suggestion
    PowerMetric{<:Metric} <: Metric
```

File: src/manifolds/PowerManifold.jl [hunk 2]
```diff
@@ -82,13 +82,17 @@ function PowerManifold(
...
 
 Represent the [`Metric`](@ref) on an [`AbstractPowerManifold`](@ref), i.e. the inner
 product on the tangent space is the sum of the inner product of each elements
 tangent space of the power manifold.
+
+It is constructed based on the given `metric` on the powered space.
```
Comment (kellertuer) [inline src/manifolds/PowerManifold.jl:91]: Could you the explicit constructor signature here?

File: src/manifolds/ProductManifold.jl [hunk 3]
```diff
@@ -42,12 +42,22 @@ end
 
 const PRODUCT_BASIS_LIST_CACHED = [CachedBasis]
 
-"""
+@doc raw"""
     ProductMetric <: Metric
```
Comment (kellertuer) [inline src/manifolds/ProductManifold.jl:46]:
```suggestion
    ProductMetric{<:(Tuple{N,Metric} where N)} <: Metric
```

File: src/manifolds/PowerManifold.jl (hunk 2 above)
Comment (mateuszbaran) [reply src/manifolds/PowerManifold.jl]: OK, done.

File: src/manifolds/ProductManifold.jl [hunk 4]
```diff
@@ -42,12 +42,22 @@ end
 
 const PRODUCT_BASIS_LIST_CACHED = [CachedBasis]
 
-"""
-    ProductMetric <: Metric
+@doc raw"""
+    ProductMetric{<:(Tuple{N,Metric} where N)} <: Metric
+
+A type to represent the product of `metrics` for a [`ProductManifold`](@ref).
```
Comment (sethaxen) [inline src/manifolds/ProductManifold.jl:48]: `metrics` here isn't a variable or type in this docstring, so it doesn't need the ticks
```suggestion
A type to represent the product of [`Metric`](@ref)s for a [`ProductManifold`](@ref).
```

File: src/manifolds/ProductManifold.jl [hunk 5]
```diff
@@ -42,12 +42,22 @@ end
 
 const PRODUCT_BASIS_LIST_CACHED = [CachedBasis]
 
-"""
-    ProductMetric <: Metric
+@doc raw"""
+    ProductMetric{<:(Tuple{N,Metric} where N)} <: Metric
```
Comment (sethaxen) [inline src/manifolds/ProductManifold.jl:46]: The parens aren't necessary

```suggestion
    ProductMetric{<:Tuple{N,Metric} where N} <: Metric
```

**Comment by sethaxen** [comment]
Oh, also this is technically a breaking change, so we'll want to increment the version number accordingly after this. I'd like to get the distribution stuff in its own patch release, so I'll do that first.

**Comment by mateuszbaran** [comment]
Sure, I can definitely wait with this and call this 0.2.0. By the way, now that I've done this, extending this power/product metrics to all metric-dependent functions (exp, log, inner etc.) looks like something that should also be done.

**Comment by sethaxen** [comment]
> Sure, I can definitely wait with this and call this 0.2.0. By the way, now that I've done this, extending this power/product metrics to all metric-dependent functions (exp, log, inner etc.) looks like something that should also be done.

Hmm, is there a way we can use the default metric functionality to make this happen with just a few lines of code?

**Comment by mateuszbaran** [comment]
> > Sure, I can definitely wait with this and call this 0.2.0. By the way, now that I've done this, extending this power/product metrics to all metric-dependent functions (exp, log, inner etc.) looks like something that should also be done.
> 
> Hmm, is there a way we can use the default metric functionality to make this happen with just a few lines of code?

As far as I'm aware, not without some changes to the default metric functionality itself. The more I think about it, the less I like this re-wrapping that needs to be done here. And it doesn't even really add new functionality, it's just another way to achieve something that is and will be possible anyway.

**Comment by sethaxen** [comment]
> As far as I'm aware, not without some changes to the default metric functionality itself. The more I think about it, the less I like this re-wrapping that needs to be done here. And it doesn't even really add new functionality, it's just another way to achieve something that is and will be possible anyway.

I guess the main purpose is to enable one to equip a `ProductManifold` with a different product metric than the one that results from each of the (maybe default) metrics of the submanifold. That is, to avoid enforcing that if one wants to do this, that they need to equip each submanifold with a metric before taking the cartesian product. I like removing the order constraint, but since product metrics only make sense when they map one-to-one to each submanifold, it may after all be cleaner to enforce the order.

**Comment by mateuszbaran** [comment]
Honestly, it feels like too much hassle for a minor convenience. I think I'll close the PR unless you or Ronny want to work on it.

**Comment by sethaxen** [comment]
Sounds good to me.

**Comment by kellertuer** [comment]
Maybe the question is: Do we need both orders (Product wraps metric and Metric wraps product) or is one enough? Both `ProductMetric` and `PowerMetric` aim to do the second approach (and I think that requires storing the metric internally).

Is it worth it? I don't know. In the beginning I thought, since you would store the metric just once (instead of within every product factor, for power you actually really only store it once within the  (metric) manifold). So maybe the second way (MetricManifold(Power)) is not the way we want to go, but just provide (Power(MetricManifold. I would be fine with that, though that would close this and #125 with “No, we don't do that”. I would be fine with that decision, too.

**Comment by mateuszbaran** [comment]
OK, so then I'll close both this and #125 .

**Comment by kellertuer** [comment]
Yes. I think, we can even remove `PowerMetric` and `ProductMetric` then?

**Comment by mateuszbaran** [comment]
All these two metrics do is allowing people to wrap power and product manifolds in `MetricManifold`. I don't know how useful it is when it doesn't change the metric at all.

**Comment by kellertuer** [comment]
Thats exactly my conclusion, I don't think its useful, even because you can't change the metric in the current form.

**Comment by mateuszbaran** [comment]
OK, so we can remove it then.

**Comment by sethaxen** [comment]
Last I checked some metric functions only worked for metric manifolds. If
that's still the case, then these metrics would still be necessary.

**Comment by kellertuer** [comment]
No, you would put the metric on the one (for power) or the element-manifolds (for product) and then it works again, I think? But just to be sure we can also leave them in for ow and check more precisely wether they are useful (if you can't change the metric).

**Pull request closed without merging**